src/util/proxyDevices.tsx

```diff
@@ -0,0 +1,131 @@
+import { Input, Label, Select } from "@canonical/react-components";
+import { MainTableRow } from "@canonical/react-components/dist/components/MainTable/MainTable";
+import { getConfigurationRowBase } from "components/ConfigurationRow";
+import { LxdProxyDevice } from "types/device";
+import { ensureEditMode } from "./instanceEdit";
+import { proxyAddressTypeOptions } from "./instanceOptions";
+import { InstanceAndProfileFormikProps } from "components/forms/instanceAndProfileFormValues";
+
+type ConnectionType = "listen" | "connect";
+
+export const getProxyAddress = (
+  customRows: MainTableRow[],
+  device: LxdProxyDevice,
+  index: number,
+  connectionType: ConnectionType,
+  formik: InstanceAndProfileFormikProps,
+) => {
+  const deviceParts = device[connectionType]?.split(":") || [];
+  const deviceType = deviceParts.length > 0 ? deviceParts[0] : "tcp";
+  const deviceAddress = deviceParts.length > 1 ? deviceParts[1] : "";
+  const devicePort = deviceParts.length > 2 ? deviceParts[2] : "";
+
+  customRows.push(
+    getConfigurationRowBase({
+      className: "no-border-top inherited-with-form",
+      configuration: <Label forId={`devices.${index}.type`}>Type</Label>,
+      inherited: (
+        <Select
+          onChange={(e) => {
+            ensureEditMode(formik);
+            const newType = e.target.value;
+            void formik.setFieldValue(
+              `devices.${index}.${connectionType}`,
+              `${newType}:${deviceAddress}:${devicePort}`,
+            );
+
+            if (device.nat === "true") {
+              void formik.setFieldValue(
+                `devices.${index}.connect`,
+                `${newType}:${deviceAddress}:${devicePort}`,
+              );
+            }
+          }}
+          value={deviceType}
+          options={proxyAddressTypeOptions}
+          className="u-no-margin--bottom"
+          disabled={connectionType === "connect" && device.nat === "true"}
+          title={
+            device.nat
+              ? "This is determined by the listen type when nat mode is enabled"
+              : undefined
+          }
+        />
+      ),
+      override: "",
+    }),
+  );
+
+  customRows.push(
+    getConfigurationRowBase({
+      className: "no-border-top inherited-with-form",
+      configuration:
+        deviceType === "unix" ? (
+          <Label forId={`devices.${index}.unixsocket`}>*Socket path</Label>
+        ) : (
+          <Label forId={`devices.${index}.address`}>*Address</Label>
+        ),
+      inherited:
+        deviceType === "unix" ? (
+          <Input
+            onChange={(e) => {
+              ensureEditMode(formik);
+              const newAddress = e.target.value;
+              void formik.setFieldValue(
+                `devices.${index}.${connectionType}`,
+                `${newAddress}`,
```

Comment:
We are missing to prepend with type: here, no?

Reply:
Prepend?

Reply:
Yes. This is for the unix type, right? If so we could just hardcode it as `unix:${newAddress}`

Reply:
Maybe also rename newAddress to socketPath

Reply:
Oh right. Yes that is true. I definitely missed needing to do that.